File: Python/Stack_linked_list/stack.py
class Node:
  def __init__(self, value):
    self.value = value
    self.next = None

class LinkedListStack:
  def __init__(self, head=None):
    self.head = head

  def push(self, value):
    # If list is not empty
    if self.head:
      new_node = Node(value)
      # Reference to previous head
      new_node.next = self.head
      # Replace previous head
      self.head = new_node
    else:
      # Create the first node if none
      self.head = Node(value)

  def pop(self):
    # If list is not empty
    if not self.head:
      return None
    popped = self.head.value
    self.head = self.head.next
    return popped

File: C/Stack_linked_list/stack.c
#include "stack.h"
#include <stdlib.h>
#include <stdio.h>

Node *Node_Init(short value) {
	Node *n = malloc(sizeof(struct Node));
	n->value = value;
	n->next = NULL;
	return n;
}

Stack *Stack_Init() {
	Stack *s = malloc(sizeof(struct Stack));
	s->head = NULL;
	return s;
}

void push(Stack *s, short value) {
	Node *n = Node_Init(value);
	// Reference to the previous head
	n->next = s->head;
	// Replace head
	s->head = n;
}

short pop(Stack *s) {
	if (!s->head) {
		printf("Stack is empty\n");
		return -1;
	}
	// Save value before freeing node
	Node *prevHead = s->head;
	short value = prevHead->value;
	// Change head to next node
	s->head = prevHead->next;
	free(prevHead);
	return value;
}
